File: db-migrate-dotnet/Services/SqliteReader.cs
using Microsoft.Data.Sqlite;

namespace Fuxa.DbMigrate.Services;

/// <summary>
/// Reads data from FUXA SQLite databases using Microsoft.Data.Sqlite
/// </summary>
public class SqliteReader : IDisposable
{
    private readonly string _appdataDir;
    private readonly Dictionary<string, SqliteConnection> _connections = [];
    private bool _disposed;

    public SqliteReader(string appdataDir)
    {
        _appdataDir = appdataDir;
    }

    /// <summary>
    /// Open (or reuse) a connection to a SQLite database file
    /// </summary>
    private SqliteConnection? Open(string dbFile)
    {
        var dbPath = Path.GetFullPath(Path.Combine(_appdataDir, dbFile));
        if (!File.Exists(dbPath))
            return null;

        if (_connections.TryGetValue(dbPath, out var existing))
            return existing;

        var conn = new SqliteConnection($"Data Source={dbPath};Mode=ReadOnly");
        conn.Open();
        _connections[dbPath] = conn;
        return conn;
    }

    /// <summary>
    /// Check if database file exists
    /// </summary>
    public bool Exists(string dbFile)
    {
        var dbPath = Path.GetFullPath(Path.Combine(_appdataDir, dbFile));
        return File.Exists(dbPath);
    }

    /// <summary>
    /// Get table names in a database
    /// </summary>
    public List<string> GetTableNames(string dbFile)
    {
        var conn = Open(dbFile);
        if (conn == null) return [];

        using var cmd = conn.CreateCommand();
        cmd.CommandText = "SELECT name FROM sqlite_master WHERE type='table' ORDER BY name";
        var names = new List<string>();
        using var reader = cmd.ExecuteReader();
        while (reader.Read())
        {
            names.Add(reader.GetString(0));
        }
        return names;
    }

    /// <summary>
    /// Count rows in a table
    /// </summary>
    public long Count(string dbFile, string tableName)
    {
        var conn = Open(dbFile);
        if (conn == null) return 0;

        try
        {
            using var cmd = conn.CreateCommand();
            cmd.CommandText = $"SELECT COUNT(*) FROM [{tableName}]";
            var result = cmd.ExecuteScalar();
            return result != null ? Convert.ToInt64(result) : 0;
        }
        catch (SqliteException ex) when (ex.Message.Contains("no such table"))
        {
            return 0;
        }
    }

    /// <summary>
    /// Read all rows from a table as a list of dictionaries
    /// </summary>
    public List<Dictionary<string, object?>> ReadAll(string dbFile, string tableName)
    {
        var conn = Open(dbFile);
        if (conn == null) return [];

        try
        {
            using var cmd = conn.CreateCommand();
            cmd.CommandText = $"SELECT * FROM [{tableName}]";
            var rows = new List<Dictionary<string, object?>>();
            using var reader = cmd.ExecuteReader();
            while (reader.Read())
            {
                var row = new Dictionary<string, object?>();
                for (int i = 0; i < reader.FieldCount; i++)
                {
                    row[reader.GetName(i)] = reader.IsDBNull(i) ? null : reader.GetValue(i);
                }
                rows.Add(row);
            }
            return rows;
        }
        catch (SqliteException ex) when (ex.Message.Contains("no such table"))
        {
            return [];
        }
    }

    /// <summary>
    /// Read rows in batches (for large DAQ datasets)
    /// </summary>
    public void ReadAllBatched(string dbFile, string tableName, int batchSize,
        Action<List<Dictionary<string, object?>>, int, long> callback)
    {
        var conn = Open(dbFile);
        if (conn == null) return;

        try
        {
            long total;
            using (var countCmd = conn.CreateCommand())
            {
                countCmd.CommandText = $"SELECT COUNT(*) FROM [{tableName}]";
                var result = countCmd.ExecuteScalar();
                total = result != null ? Convert.ToInt64(result) : 0;
            }

            int offset = 0;
            while (offset < total)
            {
                using var cmd = conn.CreateCommand();
                cmd.CommandText = $"SELECT * FROM [{tableName}] LIMIT @limit OFFSET @offset";
                cmd.Parameters.AddWithValue("@limit", batchSize);
                cmd.Parameters.AddWithValue("@offset", offset);

                var batch = new List<Dictionary<string, object?>>();
                using var reader = cmd.ExecuteReader();
                while (reader.Read())
                {
                    var row = new Dictionary<string, object?>();
                    for (int i = 0; i < reader.FieldCount; i++)
                    {
                        row[reader.GetName(i)] = reader.IsDBNull(i) ? null : reader.GetValue(i);
                    }
                    batch.Add(row);
                }

                if (batch.Count == 0) break;
                callback(batch, offset, total);
                offset += batchSize;
            }
        }
        catch (SqliteException ex) when (!ex.Message.Contains("no such table"))
        {
            throw;
        }
    }

    /// <summary>
    /// Find DAQ database files by prefix
    /// </summary>
    public List<string> FindDaqFiles(string prefix)
    {
        var result = new List<string>();

        try
        {
            foreach (var file in Directory.GetFiles(_appdataDir, $"{prefix}*.db"))
            {
                result.Add(Path.GetFileName(file));
            }

            // Also check archive folder
            var archiveDir = Path.Combine(_appdataDir, "archive");
            if (Directory.Exists(archiveDir))
            {
                foreach (var file in Directory.GetFiles(archiveDir, $"{prefix}*.db"))
                {
                    result.Add(Path.Combine("archive", Path.GetFileName(file)));
                }
            }
        }
        catch
        {
            // ignore
        }

        return result;
    }

    public void Dispose()
    {
        if (_disposed) return;
        _disposed = true;

        foreach (var conn in _connections.Values)
        {
            try { conn.Close(); conn.Dispose(); } catch { /* ignore */ }
        }
        _connections.Clear();

        GC.SuppressFinalize(this);
    }
}

File: db-migrate/src/sqlite-reader.js
/**
 * SQLite reader module
 * Reads data from FUXA SQLite databases using better-sqlite3 (synchronous, fast)
 */

'use strict';

const path = require('path');
const fs = require('fs');

let Database;
try {
  Database = require('better-sqlite3');
} catch (e) {
  console.error('better-sqlite3 not installed. Run: npm install');
  process.exit(1);
}

class SqliteReader {
  /**
   * @param {string} appdataDir - FUXA _appdata directory path
   */
  constructor(appdataDir) {
    this.appdataDir = appdataDir;
    this.connections = {};
  }

  /**
   * Open a SQLite database file
   * @param {string} dbFile - database file name
   * @returns {object} better-sqlite3 Database instance
   */
  open(dbFile) {
    const dbPath = path.resolve(this.appdataDir, dbFile);
    if (!fs.existsSync(dbPath)) {
      return null;
    }
    if (!this.connections[dbPath]) {
      this.connections[dbPath] = new Database(dbPath, { readonly: true, fileMustExist: true });
    }
    return this.connections[dbPath];
  }

  /**
   * Read all rows from a table
   * @param {string} dbFile - database file name
   * @param {string} tableName - table to read
   * @returns {Array} rows
   */
  readAll(dbFile, tableName) {
    const db = this.open(dbFile);
    if (!db) {
      return [];
    }
    try {
      const rows = db.prepare(`SELECT * FROM [${tableName}]`).all();
      return rows;
    } catch (err) {
      if (err.message.includes('no such table')) {
        return [];
      }
      throw err;
    }
  }

  /**
   * Read row count from a table
   * @param {string} dbFile
   * @param {string} tableName
   * @returns {number}
   */
  count(dbFile, tableName) {
    const db = this.open(dbFile);
    if (!db) {
      return 0;
    }
    try {
      const result = db.prepare(`SELECT COUNT(*) as cnt FROM [${tableName}]`).get();
      return result ? result.cnt : 0;
    } catch (err) {
      if (err.message.includes('no such table')) {
        return 0;
      }
      throw err;
    }
  }

  /**
   * Get table names in a database
   * @param {string} dbFile
   * @returns {string[]}
   */
  getTableNames(dbFile) {
    const db = this.open(dbFile);
    if (!db) {
      return [];
    }
    const rows = db.prepare("SELECT name FROM sqlite_master WHERE type='table' ORDER BY name").all();
    return rows.map(r => r.name);
  }

  /**
   * Check if database file exists
   * @param {string} dbFile
   * @returns {boolean}
   */
  exists(dbFile) {
    return fs.existsSync(path.resolve(this.appdataDir, dbFile));
  }

  /**
   * Find DAQ database files by prefix
   * @param {string} prefix - e.g. 'daq-data_' or 'daq-map_'
   * @returns {string[]} list of file names
   */
  findDaqFiles(prefix) {
    const result = [];
    try {
      const files = fs.readdirSync(this.appdataDir);
      for (const file of files) {
        if (file.startsWith(prefix) && file.endsWith('.db')) {
          result.push(file);
        }
      }
      // Also check archive folder
      const archiveDir = path.join(this.appdataDir, 'archive');
      if (fs.existsSync(archiveDir)) {
        const archiveFiles = fs.readdirSync(archiveDir);
        for (const file of archiveFiles) {
          if (file.startsWith(prefix) && file.endsWith('.db')) {
            result.push(path.join('archive', file));
          }
        }
      }
    } catch (err) {
      // ignore
    }
    return result;
  }

  /**
   * Read all rows from a table with pagination (for large datasets)
   * @param {string} dbFile
   * @param {string} tableName
   * @param {number} batchSize
   * @param {function} callback - called with each batch of rows
   */
  readAllBatched(dbFile, tableName, batchSize, callback) {
    const db = this.open(dbFile);
    if (!db) return;

    try {
      const countResult = db.prepare(`SELECT COUNT(*) as cnt FROM [${tableName}]`).get();
      const total = countResult ? countResult.cnt : 0;

      let offset = 0;
      while (offset < total) {
        const rows = db.prepare(`SELECT * FROM [${tableName}] LIMIT ? OFFSET ?`).all(batchSize, offset);
        if (rows.length === 0) break;
        callback(rows, offset, total);
        offset += batchSize;
      }
    } catch (err) {
      if (!err.message.includes('no such table')) {
        throw err;
      }
    }
  }

  /**
   * Close all database connections
   */
  closeAll() {
    for (const dbPath of Object.keys(this.connections)) {
      try {
        this.connections[dbPath].close();
      } catch (e) {
        // ignore
      }
    }
    this.connections = {};
  }
}

module.exports = SqliteReader;
